### PR TITLE
[Frontend] Metadata cache unit tests

### PR DESCRIPTION
**Pull request by snalli**
Add unittests to evaluate metadata cache

### CODE REVIEW

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java [hunk 1]
```diff
@@ -1367,6 +1367,326 @@ public void onEvent(ReplicaId replicaId, Object e) {
...
+      // delete simple blob
+      router.deleteBlob(blobId, null).get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+      assertTrue("Must delete metadata in cache", deleteCacheEntryLatch.await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+      assertNull("Blob metadata must be absent in metadata cache", router.getBlobMetadataCache().getObject(blobId));
+    } finally {
+      router.close();
+    }
+  }
+
+  public void routerMetadataCacheErrorOnGetCompositeBlob(ServerErrorCode serverErrorCode, RouterErrorCode routerErrorCode, boolean deleteCacheEntryOnFail) throws Exception {
```
Comment (justinlin-linkedin) [inline ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1542]: make this method private please.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java [hunk 2]
```diff
@@ -1367,6 +1367,326 @@ public void onEvent(ReplicaId replicaId, Object e) {
...
+  @Test
+  public void testRouterMetadataCacheBlobExpiredOnGetCompositeBlob() throws Exception {
+    routerMetadataCacheErrorOnGetCompositeBlob(ServerErrorCode.Blob_Expired, RouterErrorCode.BlobExpired, true);
+  }
+  @Test
+  public void testRouterMetadataCacheUnexpectedInternalErrorOnGetCompositeBlob() throws Exception {
+    routerMetadataCacheErrorOnGetCompositeBlob(ServerErrorCode.Unknown_Error, RouterErrorCode.UnexpectedInternalError, false);
+  }
+
+  public void routerMetadataCacheErrorOnDeleteCompositeBlob(ServerErrorCode serverErrorCode, RouterErrorCode routerErrorCode) throws Exception {
```
Comment (justinlin-linkedin) [inline ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1625]: no need to create this function, we can just copy all the code to the test function below.

**Review by justinlin-linkedin — CHANGES_REQUESTED** [review]
The tests look good.
Can we also add a test on GetBlobOperationTest file. This file shares a lot of similarity with NonBlockingRouterTest file.
In GetBlobOperationTest file, we can add a test case to verify the GetRequest would be requesting BlobInfo instead of BlobAll when there is a cache hit in the MockNetworkClient.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java [hunk 3]
```diff
@@ -1367,6 +1367,326 @@ public void onEvent(ReplicaId replicaId, Object e) {
...
+      };
+
+      setRouter(properties, layout, loggingNotificationSystem);
+      assertNotNull("Metadata cache must be created", router.getBlobMetadataCache());
+
+      // put a simple blob and test it is absent in the metadata cache
+      setOperationParams((int) (8 * Math.pow(1024, 2)), TTL_SECS);
+      String blobId =
+          router.putBlob(putBlobProperties, putUserMetadata, putChannel, new PutBlobOptionsBuilder().build())
+              .get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
```
[the remark this answers is not in the document]
Comment (snalli) [reply ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1455]: thanks

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java (hunk 1 above)
Comment (snalli) [reply ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1542]: agreed

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java (hunk 2 above)
Comment (snalli) [reply ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1625]: yes but this would be needed if we want to test with different error combos. Right now there is just one, but I was hoping we'd add more in the future

**Comment by snalli** [comment]
> The tests look good. Can we also add a test on GetBlobOperationTest file. This file shares a lot of similarity with NonBlockingRouterTest file. In GetBlobOperationTest file, we can add a test case to verify the GetRequest would be requesting BlobInfo instead of BlobAll when there is a cache hit in the MockNetworkClient.

yes will do

File: ambry-commons/src/main/java/com/github/ambry/commons/AmbryCache.java [hunk 4]
```diff
@@ -54,12 +56,15 @@ public class AmbryCache {
    * @param cacheEnabled Toggles cache. If true, cache is enabled. Else, cache is disabled.
    * @param cacheMaxSizeBytes Maximum memory footprint of the cache
    * @param metricRegistry Instance of metrics registry to record stats
+   * @param notificationSystem
    */
-  public AmbryCache(String cacheId, boolean cacheEnabled, long cacheMaxSizeBytes, MetricRegistry metricRegistry) {
+  public AmbryCache(String cacheId, boolean cacheEnabled, long cacheMaxSizeBytes, MetricRegistry metricRegistry,
```
[the remark this answers is not in the document]
Comment (snalli) [reply ambry-commons/src/main/java/com/github/ambry/commons/AmbryCache.java:61]: On second thought, I think I found a way to do it without notifications. I'll implement and publish.

[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-commons/src/main/java/com/github/ambry/commons/AmbryCache.java:61]: The point is to not use NotificationSystem, since NotificationSystem was created for a specific reason and adding cache hit and miss event to NotificationSystem is not good idea. Let's make sure each class follow the true purpose of their creation. 
That being said, I don't oppose passing a "CacheEventCallback" or something like that, as long as it's not NotificationSystem.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java [hunk 5]
```diff
@@ -1367,6 +1365,258 @@ public void onEvent(ReplicaId replicaId, Object e) {
...
+      assertTrue("Must encounter a cache miss", ambryCacheWithMetrics.getCacheMissCountDown().await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+      assertTrue("Must put metadata in cache", ambryCacheWithMetrics.getPutObjectCountDown().await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+      assertNotNull("Blob metadata must be present in metadata cache", router.getBlobMetadataCache().getObject(blobId));
+
+      // failed delete
+      mockServerLayout.getMockServers().forEach(mockServer -> mockServer.setServerErrorForAllRequests(serverErrorCode));
+      TestCallback<Void> testCallback = new TestCallback<>();
+      Future<Void> future = router.deleteBlob(blobId, null, testCallback, null);
+      assertFailureAndCheckErrorCode(future, testCallback, routerErrorCode);
+      assertNotNull("Blob metadata must be present in metadata cache", router.getBlobMetadataCache().getObject(blobId));
```
Comment (Arun-LinkedIn) [inline ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1597]: This is not related to tests. But just wanted to note that similar to how we are invalidating cache when blob is not found during GetBlob operations, we should update logic sometime later to invalidate cache when blob is not found during GETBlobInfo, TTLUpdate or DELETE operations too. 

This will avoid BlobInfo call during later GET operations. Just wondering that even though one BlobInfo is not costly, since we already saw the blob is gone, we  should probably make use of it.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java [hunk 6]
```diff
@@ -1367,6 +1365,258 @@ public void onEvent(ReplicaId replicaId, Object e) {
...
+      setRouterWithMetadataCache(properties, new AmbryCacheMetrics(1,1,1,0));
+      AmbryCacheWithMetrics ambryCacheWithMetrics = (AmbryCacheWithMetrics) router.getBlobMetadataCache();
+      assertNotNull("Metadata cache must be created", ambryCacheWithMetrics);
+
+      // put a simple blob and test it is absent in the metadata cache
+      setOperationParams(PUT_CONTENT_SIZE, TTL_SECS);
+      blobId =
+          router.putBlob(putBlobProperties, putUserMetadata, putChannel, new PutBlobOptionsBuilder().build())
+              .get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+      assertTrue("BlobID must be version 6 blobID", BlobId.getVersion(blobId) == BlobId.BLOB_ID_V6);
```
Comment (Arun-LinkedIn) [inline ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1384]: nit: This check can be removed too since this is not related to this test.

**Review by Arun-LinkedIn — COMMENTED** [review]
Tests in NonBlockingRouterTest look good. But, like you and Justin are discussing, we should probably add unit tests in GetBlobOperationTest too to probably check cases like firstChunk is instance of CachedFirstChunk, metadata is saved correctly in Cache, operation fails when blob information is different from cache, etc.

File: ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java (hunk 5 above)
Comment (snalli) [reply ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java:1597]: got it

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java [hunk 7]
```diff
@@ -123,7 +123,7 @@ public NonBlockingRouterFactory(VerifiableProperties verifiableProperties, Clust
   public Router getRouter() {
     try {
       return new NonBlockingRouter(routerConfig, routerMetrics, networkClientFactory, notificationSystem, clusterMap,
-          kms, cryptoService, cryptoJobHandler, accountService, time, defaultPartitionClass);
+          kms, cryptoService, cryptoJobHandler, accountService, time, defaultPartitionClass, null);
```
[the remark this answers is not in the document]
Comment (snalli) [reply ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouterFactory.java:126]: good catch !

File: ambry-router/src/main/java/com/github/ambry/router/CloudRouterFactory.java [hunk 8]
```diff
@@ -145,9 +146,13 @@ public Router getRouter() throws InstantiationException {
...
           getRequestHandlerPool(verifiableProperties, clusterMap, cloudDestination, cloudConfig);
       CompositeNetworkClientFactory networkClientFactory = getCompositeNetworkClientFactory(requestHandlerPool);
+      AmbryCache blobMetadataCache = new AmbryCache(routerConfig.routerBlobMetadataCacheId,
+          routerConfig.routerBlobMetadataCacheEnabled,
+          routerConfig.routerBlobMetadataCacheMaxSizeBytes,
+          routerMetrics.getMetricRegistry());
       NonBlockingRouter router =
           new NonBlockingRouter(routerConfig, routerMetrics, networkClientFactory, notificationSystem, clusterMap, kms,
-              cryptoService, cryptoJobHandler, accountService, time, defaultPartitionClass);
+              cryptoService, cryptoJobHandler, accountService, time, defaultPartitionClass, blobMetadataCache);
```
[the remark this answers is not in the document]
Comment (snalli) [reply ambry-router/src/main/java/com/github/ambry/router/CloudRouterFactory.java:155]: good catch! this should've been in the previous file

File: ambry-router/src/test/java/com/github/ambry/router/GetBlobOperationTest.java [hunk 9]
```diff
@@ -206,7 +218,7 @@ public static List<Object[]> data() {
    * @param isBandwidthThrottlingEnabled value for
    * {@link com.github.ambry.config.QuotaConfig#bandwidthThrottlingFeatureEnabled}.
    */
-  public GetBlobOperationTest(String operationTrackerType, boolean testEncryption, boolean isBandwidthThrottlingEnabled)
+  public GetBlobOperationTest(String operationTrackerType, boolean testEncryption, boolean isBandwidthThrottlingEnabled, boolean enableMetadataCache)
```
Comment (Arun-LinkedIn) [inline ambry-router/src/test/java/com/github/ambry/router/GetBlobOperationTest.java:221]: nit: @param may be missing for enableMetadataCache

**Review by Arun-LinkedIn — APPROVED** [review]

**Review by mikeyang923 — APPROVED** [review]
lgtm
